ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/SelectionUtil.java

```diff
@@ -208,4 +216,99 @@ private static Stream<SelectExpression> resolveAllColumns(
+    return mostNestedStructFields.fields().stream()
+        .map(f -> SelectExpression.of(
+            ColumnName.of(f.name()),
+            new DereferenceExpression(
+                Optional.empty(),
+                expression,
+                f.name().toUpperCase()
+            )));
+  }
+
+  private static SqlStruct resolveStructFields(
+      final LogicalSchema parentSchema,
+      final Expression expression
+  ) {
+    // Different expressions are accepted in this method.
+    //
+    // 1. DereferenceExpression. Contains struct columns with -> references. i.e. `col0->l1`
+    // 2. ColumnReferenceExp. Contains column references only. i.e. `col0`
+    //
+    // This resolver goes back to the column reference, so it has access to the column schema. It
+    // then goes forward each struct field, specified in the DereferenceExpression, to get
+    // access to the nested fields.
+
+    // This is the column part of the parent schema, which is used to access the column type
+    if (expression instanceof ColumnReferenceExp) {
+      final ColumnName columnName = ((ColumnReferenceExp) expression).getColumnName();
+      final Optional<SqlType> parentColumnType = parentSchema
+          .findColumn(columnName)
+          .map(Column::type);
+
+      // Return the column type. If a struct, then it'll return the struct schema with all its
+      // fields. i.e. `struct<f1 int, f2 string, ...>`
+      if (parentColumnType.isPresent() && parentColumnType.get() instanceof SqlStruct) {
+        return ((SqlStruct) parentColumnType.get());
+      } else {
+        throw new IllegalArgumentException("Column " + columnName + " is not a STRUCT type.");
+      }
+    }
+
+    // This is a struct column with a reference to one of its fields. i.e. `col0->f1`
+    // In this case, it gets the base expression or struct and calls the resolver to find the
+    // parent column schema until
+    if (expression instanceof DereferenceExpression) {
+      final String fieldName = ((DereferenceExpression) expression).getFieldName();
+
+      final SqlStruct parentStruct =
+          resolveStructFields(parentSchema, ((DereferenceExpression) expression).getBase());
+
+      // Once all struct fields are obtained from the previous call, then find the field
+      // referenced in the DereferenceExpression
+      final Optional<Field> parentStructField = parentStruct.field(fieldName);
+
+      if (parentStructField.isPresent()) {
+        if (parentStructField.get().type() instanceof SqlStruct) {
+          return (SqlStruct) parentStructField.get().type();
+        } else {
+          throw new IllegalArgumentException("Field " + fieldName + " is not a STRUCT type.");
+        }
+      } else {
+        throw new IllegalArgumentException("Field " + fieldName + " was not found on STRUCT.");
+      }
+    }
+
+    throw new IllegalArgumentException(
+        "Unsupported struct column expression: " + expression.getClass().getName());
```

Comment:
The only other `Expression` I'm mildly amused by is the `Case` Expressions...  that is probably a bit of stretch though?